File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        res = [1]
        for i in range(len(nums) - 1):
            res.append(nums[i] * res[i])
        
        post = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] = res[i] * post
            post = post * nums[i]
        
        return res

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        vector<int> res;
        res.push_back(1);

        int size = nums.size();
        for (int i = 0; i < size - 1; i++) {
            res.push_back(nums[i] * res[i]);
        }

        int post = 1;
        for (int i = size - 1; i >= 0; i--) {
            res[i] *= post;
            post *= nums[i]; 
        }

        return res;
    }
};